appwrite/appwrite
# Pass custom function $data to execution

## What does this PR do?

This PR allows for custom data to be sent to ` /v1/functions/{functionId}/executions`. `$data` is sent as a string via the env `APPWRITE_FUNCTION_DATA` to the function to be decoded in whatever manner.

This also sends the `projectId`, executor's `userId` and generated JWT to the function as well, `APPWRITE_PROJECT_ID`, `APPWRITE_FUNCTION_USER_ID` and `APPWRITE_FUNCTION_JWT`, respectively.

## Test Plan

- Create new cloud function in whatever language that prints the new env vars
- POST some string `$string` in the request body with key `data` to the executions endpoint
    OR
  Execute manually, entering a custom string in the input field.
- Verify that STDOUT matches the input data, and a JWT is created

## Todo

- [x] Update POST to accept `$data`
- [x] Pass `userId` and create JWT for execution 
- [x] Create UI for adding custom data to execution
- [x] Write E2E tests

## UI Preview

![customdatastring](https://user-images.githubusercontent.com/9708641/111620700-723b7500-87bd-11eb-863f-7ca898cbf992.gif)

## Related PRs and Issues

#962 
 
### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

yes.
